### PR TITLE
Sky: Adapt radiance size if `AUTOMATIC_MODE` resolves to `REALTIME`

### LINKED ISSUE #76166: Vulkan Forward+: Visual glitch appears in the radiance map at sizes higher or lower than the default
### Godot version

4.02 stable

### System information

windows 11, GTX 1050, forward+

### Issue description

When I change the size of the radial size strange artifacts occur. If at a size of 256 the picture is expected, then at a smaller size for some reason the cubic nature of the picture becomes visible, and at a larger size incomprehensible darkened areas appear.
radial size = 32: Strange Cubic structure;
image
radial size = 64: Strange Cubic structure;
image
radial size = 128: Strange Cubic structure;
image
radial size = 256: Awesome picture;
image
radial size = 512: Strange light windows;
image
radial size = 1024: Strange light windows;
image

### Steps to reproduce

Just change radial size in sky settings;

### Minimal reproduction project

[sky.zip](https://github.com/godotengine/godot/files/11249929/sky.zip)

### PR DESCRIPTION
**Pull request by Breush**
* *Bugsquad edit, fixes: #76166*

### CODE REVIEW

**Review by Calinou — APPROVED** [review]
Tested locally, it works as expected. Code looks good to me.

**Comment by akien-mga** [comment]
For the record, your commit seems not to be linked to your GitHub account. See: [Why are my commits linked to the wrong user?](https://docs.github.com/en/pull-requests/committing-changes-to-your-project/troubleshooting-commits/why-are-my-commits-linked-to-the-wrong-user) for more info.

**Comment by Breush** [comment]
> For the record, your commit seems not to be linked to your GitHub account. See: [Why are my commits linked to the wrong user?](https://docs.github.com/en/pull-requests/committing-changes-to-your-project/troubleshooting-commits/why-are-my-commits-linked-to-the-wrong-user) for more info.

Thanks, I didn't noticed, updated my config.

File: servers/rendering/renderer_rd/environment/sky.cpp [hunk 1]
```diff
@@ -630,14 +630,39 @@ bool SkyRD::Sky::set_radiance_size(int p_radiance_size) {
 	return true;
 }
 
+RS::SkyMode SkyRD::Sky::get_effective_mode() const {
+	if (mode != RS::SKY_MODE_AUTOMATIC) {
+		return mode;
+	}
+
+	SkyMaterialData *material_data = static_cast<SkyMaterialData *>(RendererRD::MaterialStorage::get_singleton()->material_get_data(material, RendererRD::MaterialStorage::SHADER_TYPE_SKY));
```
Comment (clayjohn) [inline servers/rendering/renderer_rd/environment/sky.cpp:638]: This will fail when using the compatibility renderer since it does not use RendererRD

**Review by clayjohn — CHANGES_REQUESTED** [review]
I don't think this is the correct fix for two reasons:
1. `material_get_data()` isn't thread safe. So this change introduces a race condition when using multithreaded rendering
2. This is hardcoded to only work with the Forward+ and Mobile rendering methods. It will fail when using Compatibility. 

I think the correct fix is to not fall back to the realtime mode when the radiance size is not 256. I.e. check for radiance size here:
https://github.com/godotengine/godot/blob/61598c5c88d95b96811d386cb20d714c35f4c6d7/servers/rendering/renderer_rd/environment/sky.cpp#L1253-L1256

and here:

https://github.com/godotengine/godot/blob/61598c5c88d95b96811d386cb20d714c35f4c6d7/drivers/gles3/rasterizer_scene_gles3.cpp#L914-L917

e.g. `if ((shader_data->uses_time || shader_data->uses_position) && sky->radiance_size == 256)`

**Review by clayjohn — APPROVED** [review]
Looks great!

Note for maintainers, this should be good for cherry picking

**Comment by akien-mga** [comment]
Thanks!

**Comment by akien-mga** [comment]
Cherry-picked for 4.3.1.

**Pull request merged**